### DISCUSSION TITLE
Performace/threading issue - gunicorn workers spend most time waiting for an internal queue

### DISCUSSION THREAD

**Discussion by VoidDruid** [fastapi/fastapi#7141] [Questions]
# Gunicorn with Uvicorn workers mostly waits for thread pool queue

**TL;DR -** guvicorn+uvicorn spawn threads to handle requests, but they mostly wait for [this queue](https://github.com/python/cpython/blob/3.8/Lib/concurrent/futures/thread.py#L78) instead of working

### Description

I'm using the same scripts as in [tiangolo's example](https://github.com/tiangolo/uvicorn-gunicorn-docker) to run my app, just a bit streamlined for my workflow. All of my endpoints are sync because I'm using sqlalchemy extensively.

On my small "qa" VPS (2x2 GHz CPU), that I also use to run PostgreSQL, I manage to reach only about 300-350 rps (with locust), so I profiled the app and found out that it spends most of its time waiting for [this queue in thread pool executor](https://github.com/python/cpython/blob/3.8/Lib/concurrent/futures/thread.py#L78).
I tried web concurrency 2 and 4, but it did not matter much. I tried just adding `async` to function definitions, and while it boosted performance about 10%, the behavior was the same.

Am I missing some additional configuration, or is this normal?

### Environment

* OS: Host Ubuntu 18.04
* FastAPI Version: 0.60.2
* Python: 3.8.6
* Gunicorn: 20.0.4
* Uvicorn: 0.10.9

App runs in Docker, command: `gunicorn -k uvicorn.workers.UvicornWorker -c "$GUNICORN_CONF" "$APP_MODULE"`

### py-spy graph
Flame graph does not really change depending on which thread/process is profiled.
![profile3](https://user-images.githubusercontent.com/30983257/95302812-1e7e4480-088b-11eb-83f7-d64e6484541e.png)
Source SVG: 
[profile3.svg.zip](https://github.com/tiangolo/fastapi/files/5339026/profile3.svg.zip)

**Comment by artslob** [comment]
Can it be because your app is highly asynchronous so it mostly waits for scheduled tasks? Or app cant handle load more then 350 requests/sec?

>All of my endpoints are sync because I'm using sqlalchemy extensively

As I know sync endpoint are executed in thread pool executor and awaited using asyncio. Maybe you reach limit of available threads? Did you try to use asynchronous database (as https://github.com/encode/databases or https://python-gino.org/ for example)?

**Comment by VoidDruid** [comment]
It is too late to migrate from sqlalchemy to anything else, unfortunately, but I might rewrite it to use sqlalchemy async api when it is released. I will look into thread limits, thanks

@artslob There is no scheduled tasks inside the app itself, it is a relatively basic CRUD app for geospatial data. It does wait for database requests, but those are shown in flame graph as "do_execute". Is there a way to see if it actually waits for db, or just idle?

Database CPU usage is a fraction of what the app uses, so the workers must be doing something else than waiting for database, I assume. During load testing CPU usage is 100%.

**Comment by artslob** [comment]
The [function](https://github.com/python/cpython/blob/687af44df8caa69312b65d1bd7bf1f05dd1e5778/Lib/concurrent/futures/thread.py#L66) `_worker` you provided is [used](https://github.com/python/cpython/blob/687af44df8caa69312b65d1bd7bf1f05dd1e5778/Lib/concurrent/futures/thread.py#L207) later in `ThreadPoolExecutor` as main function of thread-worker. So I assume that this code
```python
        while True:
            work_item = work_queue.get(block=True)
```
means that thread worker waits for work to do, but there is no work. So most of time your applications just waits for work (I assume). Maybe locust cant provide high load for your application, and 350 req/sec is limit of locust?

Also, does graph shows only main gunicorn process or it shows all processes/threads?

---
>During load testing CPU usage is 100%

Do you run tests and application on same machine or on different?

PS: Im not ASGI expert, just trying to help :)

**Comment by VoidDruid** [comment]
> So most of time your applications just waits for work

That was my conclusion too, after looking at the code, but I do not understand why it does that

> Maybe locust cant provide high load for your application, and 350 req/sec is limit of locust?

That was my first assumption, but 100% cpu usage indicates it can.

> Do you run tests and application on same machine or on different?

Different machines. I did not think about it before, but maybe it actually waits for the network or connection? 🤔 

I recently tried to load test my app again, using other "lighter" endpoints and was able to reach ~450 rps. I also established a baseline on this VPS - "ping" endpoint reaches about 700 rps and a small Django app that I load-tested to eliminate the factor of FastAPI and asgi reached about 500 rps - it is all in the same neighbourhood, roughly speaking.

So it looks like either the gunicorn itself, raw cpu power or locust's measurments that is the main problem - most likely the CPU, but I'm still a bit disappointed by the performance. I will also look into the possible network problems. Closing this issue, as it does not seem to be the problem with FastAPI.

@artslob, thanks for some ideas 🙌

**Comment by artslob** [comment]
@VoidDruid did you used `HttpUser` or `FastHttpUser` for tests? FastHttpUser is [faster alternative](https://docs.locust.io/en/stable/increase-performance.html). I tried to load test ping endpoint of my service writted with FastAPI, this are results I got.

Simple ping router:
```python
router = APIRouter()


class PingModel(BaseModel):
    answer: str = "pong"
    version: str = __version__


@router.get("/ping/", response_model=PingModel)
async def pong():
    return PingModel()
```
Command to run server:
```
gunicorn -w 10 -k uvicorn.workers.UvicornWorker 'cyberbox.asgi:app'
```
Command for locust:
```
locust -f tests/load_testing/script.py --host http://127.1:8000 -u 1000 -r 10 PingUser
```
Script:
```
class PingUser(HttpUser):
    wait_time = constant_pacing(0.5)

    @task
    def ping(self):
        self.client.get("/common/ping/")
```
Results with regular user:
```
 Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /common/ping/                                             276622   223(0.08%)  |     859       1   13951     890  |  583.74    0.47
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                    276622   223(0.08%)  |     859       1   13951     890  |  583.74    0.47

Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /common/ping/                                                     890    980   1000   1100   1100   1200   1300   1600   8700  13000  14000 276622
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                        890    980   1000   1100   1100   1200   1300   1600   8700  13000  14000 276622
```
Command for locust:
```
locust -f tests/load_testing/script.py --host http://127.1:8000 -u 2000 -r 30 PingUser
```
Script:
```
from locust.contrib.fasthttp import FastHttpUser


class PingUser(FastHttpUser):
    wait_time = constant_pacing(0.5)

    @task
    def ping(self):
        self.client.get("/common/ping/")
```
Results with fast client:
```
 Name                                                          # reqs      # fails  |     Avg     Min     Max  Median  |   req/s failures/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /common/ping/                                            2435172     0(0.00%)  |     136       0  104875      77  | 2060.35    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregated                                                   2435172     0(0.00%)  |     136       0  104875      77  | 2060.35    0.00

Response time percentiles (approximated)
 Type     Name                                                              50%    66%    75%    80%    90%    95%    98%    99%  99.9% 99.99%   100% # reqs
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 GET      /common/ping/                                                      77    100    120    130    170    200    240    270   2100  88000 105000 2435172
--------|------------------------------------------------------------|---------|------|------|------|------|------|------|------|------|------|------|------|
 None     Aggregated                                                         77    100    120    130    170    200    240    270   2100  88000 105000 2435172
```
And still it look like FastAPI can handle more req/sec even with fast locust client. Maybe it should be tried with distributed locust client.

**Comment by tiangolo** [comment]
Thanks for the help @artslob ! 🍰 

And thanks for coming back to close it @VoidDruid ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
